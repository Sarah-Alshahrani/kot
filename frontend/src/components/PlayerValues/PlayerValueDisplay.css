.player_details {
    justify-content: center;
    display: flex;
    flex-direction: row;
    border-width: thin
}