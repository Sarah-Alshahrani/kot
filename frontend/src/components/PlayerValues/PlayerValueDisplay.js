import React from "react";
import "./PlayerValueDisplay.css";
import PlayerValues from "./PlayerValues";
import GameInstance from "../../services/gameService";

import MT_RESPONSE_PLAYERS_STATUS_UPDATE from "../../services/config";

class PlayerValueDisplay extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      username: props.currentUser,
      gameRoom: props.currentRoom,
      data: props.data,
      justShowSelf: props.displayOnlySelf
    };

    GameInstance.addCallback(
      MT_RESPONSE_PLAYERS_STATUS_UPDATE,
      this.playerUpdateHandler.bind(this)
    );
  }

  playerUpdateHandler(message) {
    const content = message.content;

    if (content === undefined || content === "") return;

    let player_status = [];
    try {
      player_status = JSON.parse(content);
    } catch (e) {}
    this.setState({ data: player_status });
  }

  render() {
    if (this.state.data) {
      let you = this.state.username;
      let justSelf = this.state.justShowSelf;
      let items = [];
      this.state.data.forEach(element => {
        if (justSelf && element.username === you) {
          items.push(element);
        } else if (!justSelf && element.username !== you) {
          items.push(element);
        }
      });

      if (items && items.length) {
        let tisYou = this.state.username;

        return (
          <container className="player_details">
            {items.map((entry, index) => (
              <PlayerValues
                key={index}
                username={entry.username}
                thisIsYou={entry.username === tisYou}
                victory_points={entry.victory_points} // 0 to 10
                current_health={entry.current_health} // 0 to 10
                energy={entry.energy} // 0 or more
                location={entry.location} // 'In' or 'Out'
                cards={entry.cards}
              />
            ))}
          </container>
        );
      }
    }

    return <div>Waiting on Players</div>;
  }
}
export default PlayerValueDisplay;
