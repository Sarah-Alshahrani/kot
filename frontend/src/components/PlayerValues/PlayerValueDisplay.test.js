import React from "react";
import { shallow } from "enzyme";

import PlayerValueDisplay from "./PlayerValueDisplay";

describe("Verify PlayerValueDisplay layout template", () => {
  const test_player_details = [
    {
      username: "howdy",
      victory_points: 1,
      health: 0,
      energy: 2,
      position: "Out"
    },
    {
      username: "ya",
      victory_points: 5,
      health: 5,
      energy: 4,
      position: "In"
    },
    {
      username: "all",
      victory_points: 10,
      health: 10,
      energy: 6,
      position: "Out"
    }
  ];

  const test_username = "test_user";
  const test_roomname = "test_room";

  const expected_state_values = {
    username: test_username,
    gameRoom: test_roomname,
    data: test_player_details
  };

  it("Verify parameter provided PlayeValues are rendered correctly", () => {
    const component = shallow(
      <PlayerValueDisplay
        currentUser={test_username}
        currentRoom={test_roomname}
        data={test_player_details}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("Verify state transfers via props as expected", () => {
    const wrapper = shallow(
      <PlayerValueDisplay
        currentUser={test_username}
        currentRoom={test_roomname}
        data={test_player_details}
      />
    );
    expect(wrapper.state()).toEqual({ ...expected_state_values }); // passed
  });

  it("playerUpdateHandler method updates state as expected", () => {
    const wrapper = shallow(
      <PlayerValueDisplay
        currentUser={test_username}
        currentRoom={test_roomname}
        data={[]}
      />
    );

    const expected_state_values_with_no_player_details = {
      username: test_username,
      gameRoom: test_roomname,
      data: []
    };

    expect(wrapper.state()).toEqual({
      ...expected_state_values_with_no_player_details
    });

    const expected_backend_callback_message = {
      user: test_username,
      room: test_roomname,
      content: JSON.stringify(test_player_details)
    };

    wrapper.instance().playerUpdateHandler(expected_backend_callback_message);

    expect(wrapper.state()).toEqual({ ...expected_state_values }); // passed
  });
});
