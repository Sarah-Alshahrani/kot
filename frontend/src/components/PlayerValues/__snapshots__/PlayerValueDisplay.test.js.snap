// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify PlayerValueDisplay layout template Verify parameter provided PlayeValues are rendered correctly 1`] = `
<container
  className="card"
>
  <PlayerValues
    energy={2}
    health={0}
    key="0"
    player_name="howdy"
    victory_points={1}
  />
  <PlayerValues
    energy={4}
    health={5}
    key="1"
    player_name="ya"
    victory_points={5}
  />
  <PlayerValues
    energy={6}
    health={10}
    key="2"
    player_name="all"
    victory_points={10}
  />
</container>
`;
